go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,100 @@
+package libovsdbops
+
+import (
+	"fmt"
+)
+
+const (
+	// db ownership ExternalIDs
+	// when possible, use DbIndex and function from this file instead of directly using the following ExternalIDs
+	OwnerController = "owner_controller"
+	OwnerObjectType = "owner_object_type"
+	OwnerNamespace  = "owner_namespace"
+	OwnerName       = "owner_name"
+	InternalID      = "internal_id"
+)
+
+type DbIndex struct {
+	// OwnerController specifies which controller owns the object
+	// Controller should only change objects it owns, make sure to always set this field when searching with predicate
+	OwnerController string
+	// OwnerObjectType should identify object type that owns db object, e.g. Pod, NetworkPolicy, etc.
+	// Controller handlers should use unique OwnerTypes to only change objects they manage.
+	OwnerObjectType string
+	// OwnerNamespace and OwnerName should provide unique index for objects owned by the same OwnerObjectType,
+	// e.g. if there is only 1 object owned by OwnerObjectType in the system, both Namespace and Name may be empty.
+	OwnerNamespace string
+	OwnerName      string
+	// InternalID may be used to additionally identify db objects owned by same OwnerNamespace and OwnerName.
+	InternalID string
+}
+
+func GetDbIndexFromExternalIDs(externalIDs map[string]string) *DbIndex {
+	return &DbIndex{
+		externalIDs[OwnerController],
+		externalIDs[OwnerObjectType],
+		externalIDs[OwnerNamespace],
+		externalIDs[OwnerName],
+		externalIDs[InternalID],
+	}
+}
+
+func (dbIndex *DbIndex) GetExternalIDs() map[string]string {
+	externalIDs := map[string]string{
+		OwnerController: dbIndex.OwnerController,
+		OwnerObjectType: dbIndex.OwnerObjectType,
+	}
+	if dbIndex.OwnerNamespace != "" {
+		externalIDs[OwnerNamespace] = dbIndex.OwnerNamespace
+	}
+	if dbIndex.OwnerName != "" {
+		externalIDs[OwnerName] = dbIndex.OwnerName
+	}
+	if dbIndex.InternalID != "" {
+		externalIDs[InternalID] = dbIndex.InternalID
+	}
+	return externalIDs
+}
+
+// GetObjName returns unique string that identifies object with given DbIndex
+func (dbIndex *DbIndex) GetObjName() string {
+	return dbIndex.OwnerController + "_" + dbIndex.OwnerObjectType + "_" + dbIndex.OwnerNamespace +
+		"_" + dbIndex.OwnerName + "_" + dbIndex.InternalID
+}
+
+// hasExternalIDsGetter should only include types that are handled by getExternalIDs
+type hasExternalIDsGetter interface {
+}
+
+func getExternalIDs(item any) map[string]string {
+	switch v := item.(type) {
```

Comment:
It would seem more natural if modelgen generated a function for every object with ExternalIDs, then you could just add GetExternalIDs as a required function for hasExternalIDsGetter. Then pass ExternalIDsGetter as a generic type to this function and get rid of the switch statement as well as get rid of having to list specific nbdb objects in the hasExternalIDsGetter interface.

Reply:
Yeah it would be easier, but I am not sure if changing modelgen code for this specific purpose makes sense,  also I would like to still list nbdb objects that are using new externalIDs in some way, to make sure these new functions are not called by accident for other objects with older externalIDs.

Reply:
I pushed a patch to https://github.com/ovn-org/libovsdb/pull/346

With this, we can just change the code to be:
```
type hasExternalIDsGetter interface {
	GetExternalIDs() map[string]string
}

func getExternalIDs(item hasExternalIDsGetter) map[string]string {
	return item.GetExternalIDs()
}

```